ucb-rit/coldfront
# Auto-approve pending project join requests after a project-specific delay

Fixes #74

**Changes**
- Replaced `BooleanField` for hard approvals for join requests with `DurationField`, which stores a time interval after which a pending request is automatically approved. The default delay is 6 hours.
- Allowed project managers to modify this setting.
    - A manager may modify the days and hours of the duration.
    - Minutes and seconds are also available, may be too fine-grained.
- Added a management command (`python manage.py auto_approve_project_join_requests`) that auto-approves all pending requests that completed their delay period.
    - This will be installed in a cron job in production.
- Added an admin view, available from the Cluster Access Requests list view, that auto-approves these requests as well.
- In the manager view for reviewing join requests, displayed the time at which each request will be auto-approved.
    - NOTE: Currently,  this is displayed in UTC.
- Added logic so that the "Add Users" button on the Project Detail view also requests cluster access.

**How to Test**
- Install new Python requirements: `pip install -r requirements.txt`.
- Run migrations: `python manage.py migrate`.
- For testing purposes, it may be convenient to lower the delay. The following sets the delay to 1 minute for all projects:
   ```
   from coldfront.core.project.models import *
   from datetime import timedelta
   Project.objects.all().update(joins_auto_approval_delay=timedelta(minutes=1))
   ```
- Create join requests to projects that have no delay vs. some delay. The latter should generate cluster access requests automatically, while the latter require running the management command or pressing the "Refresh" button on the admin view AFTER the delay period has ended.
